src/base/utils/os.cpp

```diff
@@ -0,0 +1,390 @@
+            lpValueName[name.size()] = 0;
+        }
+
+        // Discover the size of the value
+        ::RegQueryValueExW(handle, lpValueName, NULL, &type, NULL, &cbData);
+        DWORD cBuffer = (cbData / sizeof(WCHAR)) + 1;
+        LPWSTR lpData = new WCHAR[cBuffer];
+        LONG res = ::RegQueryValueExW(handle, lpValueName, NULL, &type, (LPBYTE)lpData, &cbData);
+        if (lpValueName)
+            delete[] lpValueName;
+
+        if (res == ERROR_SUCCESS) {
+            lpData[cBuffer - 1] = 0;
+            result = QString::fromWCharArray(lpData);
+        }
+        delete[] lpData;
+
+        return result;
+    }
+
+    QString pythonSearchReg(const REG_SEARCH_TYPE type)
+    {
+        HKEY hkRoot;
+        if (type == USER)
+            hkRoot = HKEY_CURRENT_USER;
+        else
+            hkRoot = HKEY_LOCAL_MACHINE;
+
+        REGSAM samDesired = KEY_READ;
+        if (type == SYSTEM_32BIT)
+            samDesired |= KEY_WOW64_32KEY;
+        else if (type == SYSTEM_64BIT)
+            samDesired |= KEY_WOW64_64KEY;
+
+        QString path;
+        LONG res = 0;
+        HKEY hkPythonCore;
+        res = ::RegOpenKeyExW(hkRoot, L"SOFTWARE\\Python\\PythonCore", 0, samDesired, &hkPythonCore);
+
+        if (res == ERROR_SUCCESS) {
+            QStringList versions = getRegSubkeys(hkPythonCore);
+            qDebug("Python versions nb: %d", versions.size());
+            versions.sort();
+
+            bool found = false;
+            while (!found && !versions.empty()) {
+                const QString version = versions.takeLast() + "\\InstallPath";
+                LPWSTR lpSubkey = new WCHAR[version.size() + 1];
+                version.toWCharArray(lpSubkey);
+                lpSubkey[version.size()] = 0;
+
+                HKEY hkInstallPath;
+                res = ::RegOpenKeyExW(hkPythonCore, lpSubkey, 0, samDesired, &hkInstallPath);
+                delete[] lpSubkey;
+
+                if (res == ERROR_SUCCESS) {
+                    qDebug("Detected possible Python v%s location", qUtf8Printable(version));
+                    path = getRegValue(hkInstallPath);
+                    ::RegCloseKey(hkInstallPath);
+
+                    if (!path.isEmpty() && QDir(path).exists("python.exe")) {
+                        qDebug("Found python.exe at %s", qUtf8Printable(path));
+                        found = true;
+                    }
+                }
+            }
+
+            if (!found)
+                path = QString();
+
+            ::RegCloseKey(hkPythonCore);
+        }
+
+        return path;
+    }
+
+    QString getCurrentAssoc(const QString &cls)
+    {
+        return (QSettings {"HKEY_CLASSES_ROOT", QSettings::NativeFormat}).value(cls + "/Default").toString();
+    }
+
+    void setCurrentAssoc(const QString &cls, const QString &value)
+    {
+        (QSettings {"HKEY_CURRENT_USER\\Software\\Classes", QSettings::NativeFormat}).setValue(cls + "/Default", value);
```

Comment:
Why do these two functions use different registry paths? Is it true that setting a value in HKCU is reflected in HKEY_CLASSES_ROOT? Or there are two default, system-wide and per-user ones?

Reply:
`HKEY_LOCAL_MACHINE\Software\Classes` contains global values, `HKEY_CURRENT_USER\Software\Classes` contains user overrides, HKEY_CLASSES_ROOT merges it so I use it for reading (when no matter where they are actually stored).

Reply:
I see, thanks.